### ISSUE TITLE
[tracing] NPE thrown when a file is generated without traces

### ISSUE THREAD

**Issue by spoenemann** [eclipse/xtext-core#359]
Using the GeneratorNodeProcessor to serialize a generator tree without any TraceNode leads to
```
java.lang.NullPointerException
	at org.eclipse.xtext.builder.EclipseResourceFileSystemAccess2.updateTraceInformation(EclipseResourceFileSystemAccess2.java:422)
	at org.eclipse.xtext.builder.EclipseResourceFileSystemAccess2.generateFile(EclipseResourceFileSystemAccess2.java:269)
	at org.eclipse.xtext.builder.EclipseResourceFileSystemAccess2.generateFile(EclipseResourceFileSystemAccess2.java:219)
```
Because an implementation of ITraceRegionProvider must not return null when asked for its trace region.

The Javadoc of ITraceRegionProvider says it should throw a TraceNotFoundException if no trace can be found. However, this case is not considered in EclipseResourceFileSystemAccess2.

**Comment by cdietrich** [comment]
other clients like

`org.eclipse.xtext.generator.JavaIoFileSystemAccess.generateTrace(String, String, CharSequence)`

ignore this contract as well.

**Comment by cdietrich** [comment]
@szarnekow what do you think? and can we still change the contract?

**Comment by szarnekow** [comment]
I'd go for consistently throwing a TraceNotFoundException in the different implementations of ITraceRegionProvider. EclipseResourceFileSystemAccess2.updateTraceInformation could simply be adjusted to handle that contract correctly.

**Comment by andreasmuelder** [comment]
+1

**Comment by cdietrich** [comment]
@szarnekow how to handle the exception correct? can the trace file be empty? write no trace file/delte it?

**Comment by cdietrich** [comment]
Fixed in Xtext 2.14
